app/src/main/java/me/phh/sip/SipMessage.kt

```diff
@@ -0,0 +1,119 @@
+package me.phh.sip
+
+import java.io.BufferedReader
+
+
+/* type definitions */
+
+enum class SipMethod {
+	REGISTER,
+	SUBSCRIBE,
+	INVITE,
+	ACK,
+	CANCEL,
+	BYE,
+	OPTIONS,
+	MESSAGE
+}
+
+data class SipStatusCode(val code: Int)
+
+data class SipHeader(val value: String, val parameters: Map<String, String>)
+
+sealed class SipMessage()
+
+data class SipCommonMessage(
+	val firstLine: String,
+	val headers: Map<String, List<SipHeader>>,
+	// TODO: messageBody needs to handle multipart, including nested multipart,
+	// so needs to be a tree of sort? We can probably flatten it into
+	// Map<String, List<ByteArray>> where first string is content type?
+	// (second part list because there could be multiple of the same
+	// and map doesn't allow duplicates)
+	// See RFC5621 section 3
+	val body: ByteArray?,
+)
+
+data class SipRequest(
+	val method: SipMethod,
+	val message: SipCommonMessage,
+) : SipMessage()
+
+data class SipResponse(
+	val statusCode: SipStatusCode,
+	val message: SipCommonMessage,
+) : SipMessage()
+
+
+/* rfc3261 section 7 describes how headers should be formed, in particular:
+ *  - spaces around : are ignored
+ *  - multilines headers are allowed, continuation lines start with a space or tab
+ *    and should be joined with a single space
+ *  - some headers are comma separated lists, multiple occurence of these headers
+ *    is identical to having separated values on single line
+ *  - some headers have parameters split by ;, thanksfully no comma separated list
+ *    has these so there is no ambiguity vs. how they'd distribute in this case...
+ *  - headers and parameters are case-insensitive
+ */
+private val splitHeader = "^(.+?) *: *(.+)$".toRegex()
+
+fun parseHeaders(input: BufferedReader): Map<String, List<SipHeader>> =
```

Comment:
I presume that unlike me, you read the specs and you did it for a reason, but aren't usually headers just Map<String, SipHeader>? Some keys can be repeated multiple times in a header and have a meaning?

Reply:
nvm, there is `P-Associated-Uri` already

Reply:
Some headers like `Contact`, `Route` etc are allowed to be provided multiple times, and either:
```
Contact: first contact, second contact
```
or
```
Contact: first contact
Contact: second contact
```
have the same meaning (now I'm checking Concact can have parameters, so my comment about never mixing parameters and commas was wrong, but it's now clear to me the parameters belong to the comma-segment so there's no ambiguity)

... and I get multiple Route headers which are defined as a comma separated list, so we need to at least faithfully remember them all. As far as I can see it's just about repeating the same Route: headers in our messages, so the current code should be fine, but if we're not going to use it we could just build a big Route: firstone, secondone for our replies if that seems easier.
I've preferred to think the other way around and split it out if we get commas for headers we care about so users won't need to worry if it's a single one or multiple when they use it.

RFC3261 also specifies Authorization can be repeated, but I didn't see it in practice:
```
   Although not a comma-
   separated list, this header field name may be present multiple times,
   and MUST NOT be combined into a single header line using the usual
   rules described in [Section 7.3](https://datatracker.ietf.org/doc/html/rfc3261#section-7.3).
```